src/source/doc/lala.1.man
NAME
   lala(1f) - interpret matrix expressions using a shell-like interface

SYNOPSIS
    lala [expression(s)] [--storage values]|--topic TOPIC|[--help|--version]

DESCRIPTION
   lala(1) is an interactive computer program that serves as a convenient
   "laboratory" for computations involving matrices. It provides easy
   access to matrix software developed by the LINPACK and EISPACK
   projects. The capabilities range from standard tasks such as solving
   simultaneous linear equations and inverting matrices, through symmetric
   and nonsymmetric eigenvalue problems, to fairly sophisticated matrix
   tools such as the singular value decomposition.

OPTIONS
   expression(s)     if expressions are supplied they are evaluated
                     and the program terminates.
   -s,--storage NNN  number of 64-bit words to use for storage. Defaults
                     to 400000.
   -t,--topic TOPIC  display help for specified topic and exit
   -h,--help         display this help and exit
   -v,--version      output version information and exit

AUTHOR
   This is heavily based on a program from the Department of Computer
   Science, University of New Mexico, by Cleve Moler.

EXAMPLES
  Sample commands

    lala --topic manual # display manual and exit

    # Example 1: introductory usage:
    lala
    A=<1 2 3;5 4 6;7 8 9>
    b=<5;6;7>
    A*b
    b*A
    det(A)
    quit

   An explanation of Example 1:

    // For this session the <> character is the LALA prompt.
     <> A=<1 2 3;5 4 6;7 8 9>            <---  you enter this
     A     =                             <---  LALA response
         1.    2.    3.
         5.    4.    6.
         7.    8.    9.
     <> b=<5;6;7>
     b     =
         5.
         6.
         7.

     <> A*b             <--- you enter "multiply A and b"

     ANS   =            <--- LALA response
        38.
        91.
       146.

     <> b*A             <---you enter "multiply b and A"
        /--ERROR                         <--- LALA response
     INCOMPATIBLE FOR MULTIPLICATION

     <> det(A)         <--- Take the determinant of A

     ANS   =           <---LALA response

        18.

     <> quit           <--- you quit LALA

     total flops        34
     ADIOS
    // --------------------------------------

   Example 2: Simple looping and conditionals are also available

    lala
    //Eigenvalue sensitivity example. See section 8 of the Users' Guide.
    B = <3 0 7; 0 2 0; 0 0 1>
    L = <1 0 0; 2 1 0; -3 4 1>,  M = L\L'
    A = M*B/M
    A = round(A)
    <X,D> = eig(A)
    long,  diag(D),  short
    cond(X)
    X = X/diag(X(3,:)),  cond(X)
    Y = inv(X'),  Y'*A*X
    for j = 1:3, c(j) = norm(Y(:,j))*norm(X(:,j));
    C
    E = -1.e-6*Y(:,1)*X(:,1)'
    eig(A + .4*E),  eig(A + .5*E)
    r = .4;  s = .5;
    while s-r > 1.e-14, t = (r+s)/2; d = eig(A+t*E); ...
      if imag(d(1))=0, r = t; else, s = t;
    long,  t = r
    A+t*e,  eig(A+t*E)
    <X,D> = eig(A+t*E);  X = X/diag(X(3,:))
    short,  cond(X)
    // --------------------------------------

   Use the HELP command for further information.
   For example, to enter HELP on the entire manual,
   display directions for using HELP and
   place a User manual in the file "lala.userguide.txt",
   enter

    lala
    <>help manual
    continue ...
    h // show directions for using "help"
    w lala.userguide.txt
    continue ...
    q
    quit


CHANGELOG.md
## M_matrix Changelog

The intent of this changelog is to keep everyone in the loop about
what's new in the M_matrix project. It is a curated, chronologically ordered
list of notable changes including records of change such as bug fixes,
new features, changes, and relevant notifications.
<!-- ======================================================================== -->
---
**2025-18-02**  John S. Urban  <https://github.com/urbanjost>
### :red_circle: DIFF:
  + Changing linear algebra basic functions to require prefix "la%" .
    
### :green_circle: ADD:
  + Added many Fortran elemental intrinsic functions including all basic
    trig functions.
<!-- ======================================================================== -->
---
**2025-05-02**  John S. Urban  <https://github.com/urbanjost>
### :red_circle: DIFF:
  + START:END:STEP  Changed START:STEP:END syntax to START:END:STEP as is
                    used in Fortran.

### :green_circle: ADD:
  + clear command can take more than one variable name
  + reshape(A) packs a matrix into a vector with one row
  + added randi()
  + added set functions from M_sets as package set%. See
    "help SET THEORY".
  + added ordering functions from M_orderpack as package order%. See
    "help ORDERING".
  + added size(), int()  function
<!-- ======================================================================== -->
---
**2025-04-22**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
  + maxloc,minloc  added maxloc and minloc.
<!-- ======================================================================== -->
---
**2025-04-18**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
  + sum            added option to sum along rows or columns, and checks on
                   parameters.

### :red_circle: DIFF:
  + rand           removed all overloading of "rand" and added randu(),
                   randn(), and randseed().

<!-- ======================================================================== -->
---
**2025-04-14**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
  +  add ^ and .^ power operators.
  + pow      raise elements of an array to a power
  + reshape  added function to reshape arrays
  + all      return 1 if all values are not zero,
             return 0 if any value is zero.
  + any      return 1 if any values are not zero,
             return 0 if all value is zero.
  + lt,le,eq,ge,gt,ne  added relational functions
  + date_and_time      added additional representations of current
                       time.
  + fmtc,fmti,fmtr     convert numeric values to string
  + pack               select array elements via a mask into a vector
  + minval,maxval      extrema of real component of an array

<!-- ======================================================================== -->
---
**2021-08-20**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: DIFF:
  +  renamed "laff" to "lala".

<!-- ======================================================================== -->
---
**2021-05-20**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: DIFF:
  +  renamed size() to shape() as it resembles the Fortran procedure of that
     name more than the Fortran size() intrinsic function..

### :green_circle: ADD:
  +  added Fortran intrinsic documentation via the "fhelp" command.

<!-- ======================================================================== -->
---
**2021-05-20**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: DIFF:
  +  renamed "mat88" to "laff".

### :green_circle: ADD:
  +  added ifexists_laff() for determining if a variable name is in use in laff(3).
  +  LAFF_PATH environment variable can be used to define a colon-delimited list of
     directories to search for exec(1) files. Names are search for as specified and
     with the ".la" suffix.

<!-- ======================================================================== -->
---
**2021-05-19**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
  +  put_into_NAMELIST() and get_into_NAMELIST() are functional, allowing most
     intrinsic types from
     scalar to vector to MxN matrix to be passed to and from the calling
     program onto the NAMELIST stack.

     Currently, returned arrays must be declared allocatable in the calling
     program, as NAMELIST() can change the size of the arrays.

<!-- ======================================================================== -->
---
**2021-05-16**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: DIFF:
  +  changed to using the full ASCII character set

### :green_circle: ADD:

  +  added getenv('varname') to allow branching based on environment
     variables

### :green_circle: FIX:

<!-- ======================================================================== -->
---
**2021-05-14**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:

  +  added delete('filename') to allow for erasing scratch files

### :green_circle: FIX:

  +  fixed "save" and "load" use of a filename with uppercase or underscore

<!-- ======================================================================== -->
---
**2021-05-08**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:

  +  all printable characters are allowed in display()
  +  underscores are allowed in variable and function names
  +  character arrays of commands may be passed to NAMELIST()
  +  began a set of unit tests
  +  add zeros() function to complement ones()

### :green_circle: DIFF:

The help text is displayed with the M_help(3f) module from an array in
memory instead of a file, simplifying usage by not requiring an external
file by default, but still allowing for easily supporting alternate help
files in future versions.

<!-- ======================================================================== -->
---
**2021-04-24**  John S. Urban  <https://github.com/urbanjost>
### :green_circle: ADD:


app/lala.f90
!==================================================================================================================================!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!==================================================================================================================================!
program bigmat
use,intrinsic :: iso_fortran_env, only : int8, int16, int32, int64
use,intrinsic :: iso_fortran_env, only : real32, real64
use,intrinsic :: iso_fortran_env, only : sp=>real32, dp=>real64
use M_matrix, only  : lala
use M_CLI2, only : set_args, lget, sget, iget, expressions=>unnamed          ! add command-line parser module
implicit none
character(len=:),allocatable :: help_text(:)
character(len=:),allocatable :: version_text(:)
integer                      :: i
integer                      :: storage
logical                      :: echo
logical                      :: markdown
   call setup()
   ! define command arguments,default values and crack command line
   call set_args('lala --markdown F --noecho F --storage:s 400000',help_text,version_text )
   echo=.not.lget('noecho')
   storage=iget('storage')
   if(size(expressions).eq.0)then
      call lala(storage)                   ! CALL LALA interactively
      call lala()                          ! CALL LALA interactively
   else
      call lala(storage,echo=echo)         ! CALL LALA to initialize it and set scratch space size
      do i=1,size(expressions)
         call lala(expressions(i))         ! CALL LALA
      enddo
   endif
   stop
contains
subroutine setup()
help_text=[ CHARACTER(LEN=128) :: &
'NAME',&
'   lala(1f) - interpret matrix expressions using a shell-like interface',&
'                                                                       ',&
'SYNOPSIS                                                               ',&
'    lala [expression(s)] [--storage values]| [ --help| --version]      ',&
'                                                                       ',&
'DESCRIPTION                                                            ',&
'   lala(1) is an interactive computer program that serves as a convenient',&
'   "laboratory" for computations involving matrices. It provides easy    ',&
'   access to matrix software developed by the LINPACK and EISPACK        ',&
'   projects. The capabilities range from standard tasks such as solving  ',&
'   simultaneous linear equations and inverting matrices, through symmetric',&
'   and nonsymmetric eigenvalue problems, to fairly sophisticated matrix   ',&
'   tools such as the singular value decomposition.                        ',&
'                                                                          ',&
'OPTIONS                                                                   ',&
'   expression(s)     if expressions are supplied they are evaluated       ',&
'                     and the program terminates.                          ',&
'   -s,--storage NNN  number of 64-bit words to use for storage. Defaults  ',&
'                     to 400000.                                           ',&
'   -h,--help         display this help and exit                           ',&
'   -v,--version      output version information and exit                  ',&
'                                                                          ',&
'AUTHOR                                                                    ',&
'   This is heavily based on a program from the Department of Computer     ',&
'   Science, University of New Mexico, by Cleve Moler.                     ',&
'                                                                          ',&
'EXAMPLES                                                                  ',&
'  Sample commands                                                         ',&
'                                                                          ',&
'    # Example 1: introductory usage:                                      ',&
'    lala                                                                  ',&
'    A=<1 2 3;5 4 6;7 8 9>                                                 ',&
'    b=<5;6;7>                                                             ',&
'    A*b                                                                   ',&
'    b*A                                                                   ',&
'    det(A)                                                                ',&
'    quit                                                                  ',&
'                                                                          ',&
'   An explanation of Example 1:                                           ',&
'                                                                          ',&
'    // For this session the <> character is the LALA prompt.              ',&
'     <> A=<1 2 3;5 4 6;7 8 9>            <---  you enter this             ',&
'     A     =                             <---  LALA response              ',&
'         1.    2.    3.                                                   ',&
'         5.    4.    6.                                                   ',&
'         7.    8.    9.                                                   ',&
'     <> b=<5;6;7>                                                         ',&
'     b     =                                                              ',&
'         5.                                                               ',&
'         6.                                                               ',&
'         7.                                                               ',&
'                                                                          ',&
'     <> A*b             <--- you enter "multiply A and b"                 ',&
'                                                                          ',&
'     ANS   =            <--- LALA response                                ',&
'        38.                                                               ',&
'        91.                                                               ',&
'       146.                                                               ',&
'                                                                          ',&
'     <> b*A             <---you enter "multiply b and A"                  ',&
'        /--ERROR                         <--- LALA response               ',&
'     INCOMPATIBLE FOR MULTIPLICATION                                      ',&
'                                                                          ',&
'     <> det(A)         <--- Take the determinant of A                     ',&
'                                                                          ',&
'     ANS   =           <---LALA response                                  ',&
'                                                                          ',&
'        18.                                                               ',&
'                                                                          ',&
'     <> quit           <--- you quit LALA                                 ',&
'                                                                          ',&
'     total flops        34                                                ',&
'     ADIOS                                                                ',&
'    // --------------------------------------                             ',&
'                                                                          ',&
'   Example 2: Simple looping and conditionals are also available          ',&
'                                                                          ',&
'    lala                                                                  ',&
'    //Eigenvalue sensitivity example. See section 8 of the Users'' Guide. ',&
'    B = <3 0 7; 0 2 0; 0 0 1>                                             ',&
'    L = <1 0 0; 2 1 0; -3 4 1>,  M = L\L''                                ',&
'    A = M*B/M                                                             ',&
'    A = round(A)                                                          ',&
'    <X,D> = eig(A)                                                        ',&
'    long,  diag(D),  short                                                ',&
'    cond(X)                                                               ',&
'    X = X/diag(X(3,:)),  cond(X)                                          ',&
'    Y = inv(X''),  Y''*A*X                                                ',&
'    for j = 1:3, c(j) = norm(Y(:,j))*norm(X(:,j));                        ',&
'    C                                                                     ',&
'    E = -1.e-6*Y(:,1)*X(:,1)''                                            ',&
'    eig(A + .4*E),  eig(A + .5*E)                                         ',&
'    r = .4;  s = .5;                                                      ',&
'    while s-r > 1.e-14, t = (r+s)/2; d = eig(A+t*E); ...                  ',&
'      if imag(d(1))=0, r = t; else, s = t;                                ',&
'    long,  t = r                                                          ',&
'    A+t*e,  eig(A+t*E)                                                    ',&
'    <X,D> = eig(A+t*E);  X = X/diag(X(3,:))                               ',&
'    short,  cond(X)                                                       ',&
'    // --------------------------------------                             ',&
'                                                                          ',&
'   Use the HELP command for further information.                          ',&
'   For example, to enter HELP on the entire manual,                       ',&
'   display directions for using HELP and                                  ',&
'   place a User manual in the file "lala.userguide.txt",                  ',&
'   enter                                                                  ',&
'                                                                          ',&
'    lala                                                                  ',&
'    <>help manual                                                         ',&
'    continue ...                                                          ',&
'    h // show directions for using "help"                                 ',&
'    w lala.userguide.txt                                                  ',&
'    continue ...                                                          ',&
'    q                                                                     ',&
'    quit                                                                  ',&
'']
version_text=[ CHARACTER(LEN=128) :: &
'PRODUCT:        GPF (General Purpose Fortran) utilities and examples',&
'PROGRAM:        lala(1)                                             ',&
'DESCRIPTION:    interpret matrix operations using a shell-like interface',&
'VERSION:        1.1, 2025-04-14                                         ',&
'AUTHOR:         John S. Urban; heavily based on the original by Cleve Moler',&
'HOME PAGE:      http://www.urbanjost.altervista.org/index.html             ',&
'']
end subroutine setup
end program bigmat
!==================================================================================================================================!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!==================================================================================================================================!
